File: 169. Majority Element/169.go
package  main

import (
	"fmt"
	"sort"
)

func majorityElement(nums []int) int {
	mymap := make(map[int]int)
	flag := len(nums)/2

	for _,value := range nums{
		count,ok := mymap[value]
		if !(ok){
			count = 0
		}
		count++
		if count > flag{
			return value
		}
		mymap[value] = count
	}
	return 0
}

func majorityElement1(nums []int) int{
	sort.Ints(nums)
	return nums[len(nums)/2]
}

//============================================================================

func majorityElement2(nums []int) int {
    m := make(map[int]int)
	for _, v := range nums {
		m[v]++
		if m[v] > len(nums) / 2 {
			return v
		}
	}
	return 0
}


func main(){
	a := []int{2,2,1,1,1,2,2}
	fmt.Println(majorityElement(a))
}

/*
这里主要可以看到go map的使用
对于不存在的键，会使用默认值，int类型就默认为0, 这样使用++操作是可以的
value,ok = map[i]  i不存在ok为false，存在为true

切片可以导入sort包来进行排序操作
*/

File: 169. Majority Element/169.js
var majorityElement = function(nums) {
    var mymap = new Map()
    flag = nums.length/2
    for (i=0;i<nums.length;i++){
        count = mymap.get(nums[i])
        if (count == null){
            count = 1
            mymap.set(nums[i],count)
        }else{
            count++
        }
        if (count>flag){
            return nums[i]
        }
        mymap.set(nums[i],count)    
    }
};

var majorityElement1 = function(nums) {
    var obj = {};    
    for(var i = 0; i < nums.length; i++){
        obj[nums[i]] = obj[nums[i]] + 1 || 1;
        if(obj[nums[i]] > nums.length / 2)  return nums[i];
    }   
};

console.log(majorityElement([1,2,2]))